tower-http/src/trace.rs

```diff
@@ -0,0 +1,348 @@
+    #[inline]
+    pub fn record_headers(mut self, record_headers: bool) -> Self {
+        self.event_emitter.record_headers = record_headers;
+        self
+    }
+
+    /// Change whether the full URI should be recorded on the span or if only the path should be
+    /// recorded.
+    ///
+    /// Defaults to only recording the path no the full URI.
+    ///
+    /// Note that this method is only accessible if you're using [`DefaultEmitTracingEvents`]
+    #[inline]
+    pub fn record_full_uri(mut self, record_full_uri: bool) -> Self {
+        self.event_emitter.record_full_uri = record_full_uri;
+        self
+    }
+}
+
+impl<S, Classifier, EventEmitter> Layer<S> for TraceLayer<Classifier, EventEmitter>
+where
+    Classifier: Clone,
+    EventEmitter: Clone,
+{
+    type Service = Trace<S, Classifier, EventEmitter>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        Trace {
+            inner,
+            latency_unit: self.latency_unit,
+            classifier: self.classifier.clone(),
+            event_emitter: self.event_emitter.clone(),
+        }
+    }
+}
+
+/// Middleware that add high level [tracing] to a [`Service`].
+///
+/// [tracing]: https://crates.io/crates/tracing
+/// [`Service`]: tower_service::Service
+#[derive(Clone, Debug)]
+pub struct Trace<S, Classifier, EventEmitter> {
+    inner: S,
+    latency_unit: LatencyUnit,
+    classifier: Classifier,
+    event_emitter: EventEmitter,
+}
+
+impl<ReqBody, ResBody, S, Classifier, EventEmitter> Service<Request<ReqBody>>
+    for Trace<S, Classifier, EventEmitter>
+where
+    S: Service<Request<ReqBody>, Response = Response<ResBody>>,
+    Classifier: ClassifyResponse<ResBody, S::Error> + Clone,
+    EventEmitter: EmitTracingEvents<ReqBody, OkClass = Classifier::OkClass, ErrClass = Classifier::ErrClass>
+        + Clone,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = ResponseFuture<S::Future, Classifier, EventEmitter, ReqBody>;
+
+    #[inline]
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: Request<ReqBody>) -> Self::Future {
+        let start = Instant::now();
+
+        let span = self.event_emitter.make_span(&req);
+
+        let future = {
+            let _guard = span.enter();
+            self.event_emitter.on_request_received(&req);
+            self.inner.call(req)
+        };
+
+        ResponseFuture {
+            future,
+            span,
+            start,
+            latency_unit: self.latency_unit,
+            classifier: self.classifier.clone(),
+            event_emitter: self.event_emitter.clone(),
+            _marker: PhantomData,
+        }
+    }
+}
+
+/// The [`Future`] produced by [`Trace`] services.
+///
+/// [`Future`]: std::future::Future
+#[pin_project]
+#[derive(Debug)]
+pub struct ResponseFuture<F, Classifier, EventEmitter, ReqBody> {
```

Comment:
nit: since the classifier and event emitter parameters have defaults on the `Layer` type, they probably should here as well? this would mean moving the `ReqBody` param before the `Classifier` and `EventEmitter` params, but I think that's fine...